File: test/lib/shortbread/anchoring_test.rb
# frozen_string_literal: true

require "test_helper"

class AnchoringTest < ActiveSupport::TestCase
  DOCUMENT = <<~TEXT.strip
    Deploys are gated on review.
    Every change ships behind a flag.
    Retries are capped at three attempts.
  TEXT

  test "capture records the quote, both context sides, and the structural position" do
    anchor = capture_quote(DOCUMENT, "ships behind a flag")

    assert_equal "ships behind a flag", anchor.quote
    assert_equal DOCUMENT.index("ships behind a flag"), anchor.start_offset
    assert anchor.prefix.end_with?("Every change ")
    assert anchor.suffix.start_with?(".")
    assert_equal 1, anchor.block_index
  end

  test "an unchanged document resolves exact at full confidence" do
    anchor = capture_quote(DOCUMENT, "Retries are capped")
    resolution = Shortbread::Anchoring.resolve(anchor, DOCUMENT)

    assert_equal :exact, resolution.status
    assert_equal DOCUMENT.index("Retries are capped"), resolution.start_offset
    assert_in_delta 1.0, resolution.confidence, 0.001
  end

  test "an edit to an earlier passage shifts the anchor and resolves moved" do
    anchor = capture_quote(DOCUMENT, "Retries are capped at three attempts")
    edited = DOCUMENT.sub("Deploys are gated on review.", "Deploys are gated on review by two people.")

    resolution = Shortbread::Anchoring.resolve(anchor, edited)

    assert_equal :moved, resolution.status
    assert_equal edited.index("Retries are capped at three attempts"), resolution.start_offset
    assert_equal "Retries are capped at three attempts", edited[resolution.start_offset...resolution.end_offset]
  end

  # The defect this product exists to avoid: the prototype's first cut silently relocated a
  # Comment from deleted text onto an identical sentence elsewhere, reporting `moved` at 0.50.
  test "a Comment on deleted text orphans rather than relocating onto an identical string elsewhere" do
    document = <<~TEXT.strip
      Section one.
      The build must stay green.
      Section two.
      The build must stay green.
      Section three.
    TEXT

    anchor = capture_quote(document, "The build must stay green", occurrence: 1)
    without_anchored_text = document.sub("Section two.\nThe build must stay green.\n", "Section two.\n")

    resolution = Shortbread::Anchoring.resolve(anchor, without_anchored_text)

    assert_equal :orphaned, resolution.status,
      "a deleted Comment was relocated onto a different section's identical sentence"
    assert_nil resolution.start_offset
    assert_in_delta 0.0, resolution.confidence, 0.001
  end

  test "a quote that no longer appears at all is orphaned" do
    anchor = capture_quote(DOCUMENT, "ships behind a flag")
    rewritten = DOCUMENT.sub("ships behind a flag", "ships behind a toggle")

    resolution = Shortbread::Anchoring.resolve(anchor, rewritten)

    assert_equal :orphaned, resolution.status
    assert_nil resolution.start_offset
  end

  test "indistinguishable candidates resolve ambiguous rather than guessing" do
    # Repeats within a single block, so context scoring ties and the structural tie-break cannot
    # separate them either. The recorded offset no longer holds the quote.
    document = "go go go go and stop."
    anchor = Shortbread::Anchoring.capture(
      source: document, start_offset: 3, length: 2, release_number: 1, path: "index.html"
    )
    relocated = "Preface line.\ngo go go and stop."

    resolution = Shortbread::Anchoring.resolve(anchor, relocated)

    assert_includes %i[ambiguous orphaned], resolution.status,
      "indistinguishable candidates were resolved to a guess"
    assert_nil resolution.start_offset
    assert_in_delta 0.0, resolution.confidence, 0.001
  end

  test "a repeated string is distinguished by its surrounding context" do
    document = <<~TEXT.strip
      Alpha notes. The build must stay green. Alpha ends.
      Beta notes. The build must stay green. Beta ends.
    TEXT

    anchor = capture_quote(document, "The build must stay green", occurrence: 1)
    shifted = document.sub("Alpha notes.", "Alpha notes, revised for clarity.")

    resolution = Shortbread::Anchoring.resolve(anchor, shifted)

    assert_includes %i[exact moved], resolution.status
    assert_equal shifted.index("Beta notes."), find_containing_line_start(shifted, resolution.start_offset),
      "context disambiguation chose the wrong occurrence"
  end

  test "offsets are counted in Unicode code points so an astral character does not desync them" do
    document = "Ship 😀 it.\nThe build must stay green."
    anchor = capture_quote(document, "The build must stay green")

    assert_equal document.index("The build must stay green"), anchor.start_offset
    assert_equal "The build must stay green",
      document[anchor.start_offset, anchor.quote.length]

    resolution = Shortbread::Anchoring.resolve(anchor, document)
    assert_equal :exact, resolution.status
  end

  test "an empty quote is orphaned rather than matching everywhere" do
    anchor = Shortbread::Anchoring.capture(
      source: DOCUMENT, start_offset: 0, length: 0, release_number: 1, path: "index.html"
    )

    assert_equal :orphaned, Shortbread::Anchoring.resolve(anchor, DOCUMENT).status
  end

  private

  def capture_quote(source, quote, occurrence: 0)
    start_offset = nth_index(source, quote, occurrence)
    Shortbread::Anchoring.capture(
      source:, start_offset:, length: quote.length, release_number: 1, path: "index.html"
    )
  end

  def nth_index(source, quote, occurrence)
    offset = source.index(quote)
    occurrence.times { offset = source.index(quote, offset + 1) }
    offset
  end

  def find_containing_line_start(source, offset)
    return nil unless offset

    source.rindex("\n", offset).then { |newline| newline ? newline + 1 : 0 }
  end
end

File: test/frontend/anchoring.test.ts
/**
 * The TypeScript half of the shared conformance matrix. Its Ruby twin at
 * test/lib/shortbread/anchoring_conformance_test.rb drives the SAME fixture file and asserts the
 * same expectations, so a disagreement between the two implementations fails a suite rather than
 * silently producing Anchors the server cannot verify.
 */

import { test } from 'node:test'
import assert from 'node:assert/strict'
import { readFileSync } from 'node:fs'
import { fileURLToPath } from 'node:url'
import { dirname, join } from 'node:path'

import { capture, resolve, type AnchorStatus } from '../../app/frontend/anchoring/anchoring.ts'

type ResolutionFixture = {
  name: string
  source: string
  quote: string
  occurrence: number
  republished: string
  status: AnchorStatus
}

const here = dirname(fileURLToPath(import.meta.url))
const fixtures = JSON.parse(readFileSync(join(here, 'conformance_fixtures.json'), 'utf8')) as {
  resolution: ResolutionFixture[]
}

const points = (text: string): string[] => Array.from(text)

/** Code-point index of the nth occurrence, matching how Ruby counts. */
const nthIndex = (source: string, quote: string, occurrence: number): number => {
  const haystack = points(source)
  const needle = points(quote)
  let seen = -1

  for (let index = 0; index <= haystack.length - needle.length; index += 1) {
    if (needle.every((character, offset) => haystack[index + offset] === character)) {
      seen += 1
      if (seen === occurrence) return index
    }
  }
  throw new Error(`quote not found: ${quote}`)
}

const anchorFor = (fixture: ResolutionFixture) =>
  capture({
    source: fixture.source,
    startOffset: nthIndex(fixture.source, fixture.quote, fixture.occurrence),
    length: points(fixture.quote).length,
    releaseNumber: 1,
    path: 'index.html',
  })

test('resolution reports the shared fixture status for every case', () => {
  for (const fixture of fixtures.resolution) {
    const resolution = resolve(anchorFor(fixture), fixture.republished)

    assert.equal(resolution.status, fixture.status, fixture.name)
  }
})

test('a resolved offset always points at the anchored quote', () => {
  for (const fixture of fixtures.resolution) {
    const anchor = anchorFor(fixture)
    const resolution = resolve(anchor, fixture.republished)
    if (resolution.startOffset === null || resolution.endOffset === null) continue

    const found = points(fixture.republished)
      .slice(resolution.startOffset, resolution.endOffset)
      .join('')
    assert.equal(found, anchor.quote, `${fixture.name}: resolved offset does not hold the quote`)
  }
})

test('an astral character counts as one code point, matching Ruby', () => {
  const source = 'Ship 😀 it.\nThe build must stay green.'
  const quote = 'The build must stay green'
  const startOffset = nthIndex(source, quote, 0)

  // Ruby reports 11 here; a UTF-16 count would report 12 and desync every later offset.
  assert.equal(startOffset, 11)
  assert.notEqual(startOffset, source.indexOf(quote))

  const anchor = capture({ source, startOffset, length: points(quote).length, releaseNumber: 1, path: 'index.html' })
  assert.equal(anchor.quote, quote)
  assert.equal(resolve(anchor, source).status, 'exact')
})

test('a Comment on deleted text orphans rather than relocating onto an identical string', () => {
  const source = 'Section one.\nThe build must stay green.\nSection two.\nThe build must stay green.\nSection three.'
  const quote = 'The build must stay green'
  const anchor = capture({
    source,
    startOffset: nthIndex(source, quote, 1),
    length: points(quote).length,
    releaseNumber: 1,
    path: 'index.html',
  })

  const resolution = resolve(anchor, source.replace('Section two.\nThe build must stay green.\n', 'Section two.\n'))

  assert.equal(resolution.status, 'orphaned')
  assert.equal(resolution.startOffset, null)
})

test('an empty quote is orphaned rather than matching everywhere', () => {
  const anchor = capture({ source: 'Anything.', startOffset: 0, length: 0, releaseNumber: 1, path: 'index.html' })

  assert.equal(resolve(anchor, 'Anything.').status, 'orphaned')
})
